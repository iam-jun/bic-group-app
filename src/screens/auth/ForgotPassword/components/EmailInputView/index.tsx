import React, { useEffect, useRef } from 'react';
import { View, StyleSheet } from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import isEmpty from 'lodash/isEmpty';

import Text from '~/baseComponents/Text';
import TextInputController from '~/beinComponents/inputs/TextInputController';
import { Button } from '~/baseComponents';
import * as validation from '~/constants/commonRegex';
import { useBaseHook } from '~/hooks';
import { IObject } from '~/interfaces/common';
import spacing from '~/theme/spacing';
import useForgotPasswordStore, { IForgotPasswordState } from '../../store';
import RequestVerifyEmailModal from '~/screens/auth/VerifyEmail/RequestVerifyEmailModal';
import { authErrors } from '~/constants/authConstants';
import showToastError from '~/store/helper/showToastError';
import useModalStore from '~/store/modal';

interface Props {
  useFormData: IObject<any>;
}

const EmailInputView: React.FC<Props> = ({ useFormData }) => {
  const theme: ExtendedTheme = useTheme();
  const { t } = useBaseHook();
  const styles = themeStyles(theme);
  const { colors } = theme;

  const actions = useForgotPasswordStore((state: IForgotPasswordState) => state.actions);
  const errorRequest = useForgotPasswordStore((state: IForgotPasswordState) => state.errorRequest);
  const loadingRequest = useForgotPasswordStore((state: IForgotPasswordState) => state.loadingRequest);
  const modalActions = useModalStore((state) => state.actions);

  const refTextInput = useRef<any>();

  const {
    getValues,
    setError,
    clearErrors,
    setValue,
    formState: { errors },
    trigger,
  } = useFormData;

  useEffect(
    () => {
      refTextInput.current?.focus();
    }, [],
  );

  useEffect(
    () => {
      if (errorRequest) {
        setError(
          'email', {
            type: 'manual',
            message: errorRequest,
          },
        );
      } else {
        clearErrors('email');
      }
    }, [errorRequest],
  );

  const checkDisableRequest = () => {
    const email = getValues('email');
    return (
      loadingRequest
      || !email
      || !isEmpty(errors?.email)
      || !validation.emailRegex.test(email)
    );
  };
  const disableRequest = checkDisableRequest();

  const handleError = (error: any) => {
    if (error?.code === authErrors.USER_NOT_FOUND_EXCEPTION) {
      const email = getValues('email');
      modalActions.showModal({
        isOpen: true,
        titleFullScreen: 'groups:group_content:btn_your_groups',
        ContentComponent: <RequestVerifyEmailModal email={email} isFromSignIn={false} />,
      });
    } else {
      if (error?.code === authErrors.LIMIT_EXCEEDED_EXCEPTION) {
        showToastError({ meta: { message: t('auth:text_err_limit_exceeded') } });
      } else {
        showToastError(error);
      }
      actions.setErrorRequest('');
    }
  };

  const onRequestForgotPassword = () => {
    const email = getValues('email');
    if (email && !disableRequest) {
      setValue(
        'code', '', { shouldValidate: false },
      );
      setValue(
        'newPassword', '', { shouldValidate: false },
      );
      setValue(
        'confirmPassword', '', { shouldValidate: false },
      );
      clearErrors(['code', 'newPassword', 'confirmPassword']);

      actions.requestResetPassword(email, handleError);
    }
  };

  const validateEmail = async () => {
    await trigger('email');
  };

  return (
    <View testID="forgot_password.require_email" style={styles.container}>
      <Text.H3 useI18n>auth:text_forgot_password</Text.H3>
      <Text.BodyS useI18n style={styles.desc}>
        auth:text_forgot_password_input_desc
      </Text.BodyS>
      <TextInputController
        useFormData={useFormData}
        name="email"
        testID="forgot_password.input_email"
        placeholder={t('auth:input_label_email')}
        placeholderTextColor={colors.neutral20}
        keyboardType="email-address"
        autoCapitalize="none"
        autoFocus
        editable={!loadingRequest}
        style={styles.inputEmailContainer}
        onSubmitEditing={onRequestForgotPassword}
        validateValue={validateEmail}
        textColor={colors.neutral60}
        ref={refTextInput}
      />
      <Button.Primary
        useI18n
        testID="forgot_password.button_send"
        disabled={disableRequest}
        loading={loadingRequest}
        onPress={onRequestForgotPassword}
        style={styles.btnSendRecoverCode}
        size="large"
      >
        auth:btn_send_recover_code
      </Button.Primary>
    </View>
  );
};

const themeStyles = (theme: ExtendedTheme) => {
  const { colors } = theme;
  return StyleSheet.create({
    container: {
      paddingTop: spacing.padding.big,
    },
    desc: {
      marginTop: spacing.margin.large,
      marginBottom: spacing.margin.large,
      color: colors.neutral80,
    },
    btnSendRecoverCode: {
      marginTop: spacing.margin.big,
    },
    inputEmailContainer: {
      marginVertical: 0,
    },
  });
};

export default EmailInputView;
