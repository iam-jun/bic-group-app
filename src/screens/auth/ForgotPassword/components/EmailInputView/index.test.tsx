import * as React from 'react';
import { Auth } from 'aws-amplify';
import { cleanup } from '@testing-library/react-native';
import EmailInputView from '.';
import {
  renderWithRedux,
  fireEvent,
} from '~/test/testUtils';
import i18n from '~/localization';

afterEach(cleanup);

describe('EmailInputView component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const useForm = {
    setError: jest.fn(),
    clearErrors: jest.fn(),
    trigger: jest.fn(),

    handleSubmit: () => jest.fn(),
    control: {
      register: jest.fn(),
      unregister: jest.fn(),
      getFieldState: jest.fn(),
      _names: {
        array: new Set('test'),
        mount: new Set('test'),
        unMount: new Set('test'),
        watch: new Set('test'),
        focus: 'test',
        watchAll: false,
      },
      _subjects: {
        watch: jest.fn(),
        array: jest.fn(),
        state: jest.fn(),
      },
      _getWatch: jest.fn(),
      _formValues: ['test'],
      _defaultValues: ['test'],
    },
    getValues: () => '',
    setValue: () => jest.fn(),
    formState: {
      errors: {
      },
    },
    watch: () => jest.fn(),
  };

  it('renders correctly', () => {
    const wrapper = renderWithRedux(
      <EmailInputView useFormData={useForm} />,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('disable button send code to email when typing an invalid email', async () => {
    const wrapper = renderWithRedux(<EmailInputView useFormData={useForm} />);

    const textInputComponent = wrapper.getByTestId('forgot_password.input_email');
    expect(textInputComponent).toBeDefined();
    fireEvent.changeText(textInputComponent, 'y');

    const buttonSend = wrapper.getByTestId('forgot_password.button_send');
    expect(buttonSend).toBeDefined();
    expect(buttonSend.props?.accessibilityState?.disabled).toBe(true);
  });

  it('request code to forgot password with valid email', async () => {
    const response = {
      code: 200,
      data: {},
      meta: {},
    };

    const spyRequestResetPassword = jest.spyOn(Auth, 'forgotPassword').mockImplementation(
      () => Promise.resolve(response) as any,
    );
    const email = 'thuquyen@tgm.vn';
    const getValues = jest.fn().mockImplementation((input: string) => {
      switch (input) {
        case 'email':
          return email;
        default:
          return '';
      }
    });
    const setValue = jest.fn();
    const clearErrors = jest.fn();
    const newUseForm = {
      ...useForm,
      getValues,
      setValue,
      clearErrors,
      formState: {
        errors: {},
      },
    };

    const wrapper = renderWithRedux(<EmailInputView useFormData={newUseForm} />);

    const textInputComponent = wrapper.getByTestId('forgot_password.input_email');
    expect(textInputComponent).toBeDefined();
    fireEvent.changeText(textInputComponent, email);

    const buttonSend = wrapper.getByTestId('forgot_password.button_send');
    expect(buttonSend).toBeDefined();
    expect(buttonSend.props?.accessibilityState?.disabled).toBe(false);

    fireEvent.press(buttonSend);

    expect(spyRequestResetPassword).toBeCalledWith(email);
  });

  it('request code to forgot password with valid email but error', () => {
    const errorMessage = i18n.t('auth:text_forgot_password_email_not_found');
    const newUseForm = {
      ...useForm,
      formState: {
        errors: {
          email: {
            type: 'manual',
            message: errorMessage,
          },
        },
      },
    };

    const wrapper = renderWithRedux(<EmailInputView useFormData={newUseForm} />);

    const buttonSend = wrapper.getByTestId('forgot_password.button_send');
    expect(buttonSend).toBeDefined();
    expect(buttonSend.props?.accessibilityState?.disabled).toBe(true);

    const errorText = wrapper.queryByTestId('text_input.text_helper');
    expect(errorText).toBeDefined();
    expect(errorText?.props?.children?.[0]).toBe(errorMessage);
  });
});
