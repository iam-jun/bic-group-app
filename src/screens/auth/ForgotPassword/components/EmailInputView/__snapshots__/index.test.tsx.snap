// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailInputView component renders correctly 1`] = `
<View
  style={
    Object {
      "paddingTop": 32,
    }
  }
  testID="forgot_password.require_email"
>
  <Text
    style={
      Array [
        Object {
          "color": "#181C32",
          "fontFamily": "BeVietnamPro-SemiBold",
          "fontSize": 20,
        },
        Object {},
        undefined,
      ]
    }
  >
    Forgot password
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#212745",
          "fontFamily": "BeVietnamPro-Light",
          "fontSize": 13,
        },
        Object {},
        Object {
          "color": "#181C32",
          "marginBottom": 16,
          "marginTop": 16,
        },
      ]
    }
  >
    Please enter your email address or mobile number to search for your account.
  </Text>
  <View
    style={
      Array [
        Object {
          "flexDirection": "column",
          "marginVertical": 4,
        },
        Object {
          "marginVertical": 0,
        },
      ]
    }
    testID="text_input"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 40,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "green",
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
                "borderWidth": 1,
                "flex": 1,
                "flexDirection": "row",
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "borderColor": "#E3E6F2",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "height": 40,
              }
            }
          >
            <TextInput
              autoCapitalize="none"
              autoFocus={true}
              editable={true}
              keyboardType="email-address"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              placeholder="Email"
              placeholderTextColor="#BAC0DE"
              selectionColor="#767D93"
              style={
                Array [
                  Object {
                    "flex": 1,
                    "fontFamily": "BeVietnamPro-Light",
                    "fontSize": 15,
                    "height": 40,
                    "paddingHorizontal": 16,
                  },
                  Object {
                    "color": "#282E53",
                  },
                ]
              }
              testID="forgot_password.input_email"
              value=""
            />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "borderRadius": 8,
        },
        Object {
          "marginTop": 32,
        },
      ]
    }
    testID="forgot_password.button_send"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingVertical": 4,
          },
          undefined,
          Object {
            "backgroundColor": "#F4EFFB",
            "height": 44,
            "paddingHorizontal": 16,
          },
        ]
      }
      testID="button.content"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#212745",
              "fontFamily": "BeVietnamPro-Medium",
              "fontSize": 15,
            },
            Object {
              "color": "#BAC0DE",
            },
            Array [
              Object {
                "textAlign": "center",
              },
              undefined,
            ],
          ]
        }
        testID="button.text"
      >
        Send code
      </Text>
    </View>
  </View>
</View>
`;
